[Security][DX] Added SecurityPasswordType

pyrech opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #... <!-- prefix each issue number with "Fix #", if any -->
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->

Few days ago, I discussed with @lyrixx about how to simplify user password handling in EasyAdmin. Because currently if we want to make user's password administrable we have to create our own custom controller to interact with `UserPasswordEncoder`.  For better DX we concluded that it would be better to add a native `SecurityPasswordType` in Symfony. So here we go:

- This new form type inherits from `PasswordType`.
- It automatically detects the user object from the parent form. This can be overridden with the option `security_user`.
- If no data is submitted, the old password is kept.

I will see to add tests if the idea is accepted.

pyrech committed:
Add SecurityPasswordType

wouterj commented on src/Symfony/Bundle/SecurityBundle/Form/SecurityPasswordType.php:
```diff
@@ -0,0 +1,77 @@
+
+    public function buildView(FormView $view, FormInterface $form, array $options)
+    {
+        $view->vars['required'] = false;
```
I would remove this. If this type is e.g. used in a registration form, I can imagine the field being required. I would leave this to the user of this type.

pyrech replied:
Yep, I don't like it so much neither. I would have preferred to override the option to just make it false by default, because it felt strange to me to require the password on edition. But could not find a way to override the default :confused:

wouterj replied:
You can implement the `public function configureOptions(OptionsResolver $resolver)` method and call `$resolver->setDefault('required', false)` to set the default to `false`.

I'm not sure however if this is expected. I would keep the behavior the same as in `PasswordType`. If a user wants to make this optional, they can always use this as `->add('password', SecurityPasswordType::class, ['required' => false])`

pyrech replied:
I tried to configure the default option but it did not work (maybe due to the inheritance of form, or another cause I don't understand 😅

But anyway, let's just remove it :+1:

wouterj commented on src/Symfony/Bundle/SecurityBundle/Form/SecurityPasswordType.php:
```diff
@@ -0,0 +1,77 @@
+{
+    private $passwordEncoder;
+
+    public function __construct(UserPasswordEncoderInterface $passwordEncoder)
```
If you inject `EncoderFactoryInterface` here, you can allow the `security_user` to also be a FQCN string (instead of the object). This makes it usefull for forms where the user object does not yet exists (e.g. a registration form).

wouterj reviewed (commented):
I like this idea! I've left two minor comments.

pyrech committed:
Remove required option override

lyrixx commented:
I would move everything in the Form component. We have already a system for extension

* https://github.com/symfony/symfony/tree/master/src/Symfony/Component/Form/Extension
* https://github.com/symfony/symfony/tree/master/src/Symfony/Component/Form/DependencyInjection

pyrech committed:
Move to Form component

pyrech committed:
Add tests

pyrech commented:
I moved the form type to the form component in a new `SecurityExtension` :+1:

I also added some tests. I'm unsure if I need to add a dev-dependency on `symfony/security` inside `symfony/form` component and if yes, on which version : 5.1 but not released, will this break the CI?

lyrixx commented:
> I also added some tests. I'm unsure if I need to add a dev-dependency on `symfony/security` inside `symfony/form` component and if yes, on which version : 5.1 but not released, will this break the CI?

Yes :) IHMO you can target 3.4 because it works with 3.4 :)

yceruto commented on src/Symfony/Component/Form/Extension/Security/Type/SecurityPasswordType.php:
```diff
@@ -0,0 +1,70 @@
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setDefault('security_user', null);
+        $resolver->setAllowedTypes('security_user', [
```
the statement should be in one line

yceruto commented on src/Symfony/Component/Form/Extension/Security/Type/SecurityPasswordType.php:
```diff
@@ -0,0 +1,70 @@
+        $resolver->setAllowedTypes('security_user', [
+            'null',
+            UserInterface::class,
+        ]);
```
we've recently added a fluent interface to define options, let's use it.

yceruto commented on src/Symfony/Component/Form/Extension/Security/Type/SecurityPasswordType.php:
```diff
@@ -0,0 +1,70 @@
+                    throw new InvalidConfigurationException(sprintf('You should either use "%s" inside a parent form where data is an instance of "%s" or specify the user in "security_user" option', self::class, UserInterface::class));
+                }
+
+                $securityUser = $parentData;
```
the aux variable can be avoided here. i.e. `$securityUser = $event->getForm()->getParent()->getData()`

yceruto commented on src/Symfony/Component/Form/Extension/Security/Type/SecurityPasswordType.php:
```diff
@@ -0,0 +1,70 @@
+            $securityUser = $event->getForm()->getConfig()->getOption('security_user');
+
+            if (!$securityUser) {
+                $parentData = $event->getForm()->getParent()->getData();
```
the parent form cannot be null here, should we check?

yceruto commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml:
```diff
@@ -111,5 +111,15 @@
+        <!-- SecurityExtension -->
+        <service id="form.extension.security" class="Symfony\Component\Form\Extension\Security\SecurityExtension">
+            <argument type="service" id="security.user_password_encoder" on-invalid="ignore"/>
+            <tag name="form.type_extension" />
```
The target class isn't a form type extension, it's a "form extension" which doesn't need to be registered as service.

yceruto commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml:
```diff
@@ -111,5 +111,15 @@
+            <tag name="form.type_extension" />
+        </service>
+        <service id="form.extension.security.type.security_password" class="Symfony\Component\Form\Extension\Security\Type\SecurityPasswordType">
+            <argument type="service" id="security.user_password_encoder" on-invalid="ignore"/>
```
This class would be pointless without the required password encoder, I'd suggest be strict here and remove if the service definition if security isn't enabled. See `form.type_extension.form.validator` as sample.

yceruto commented on src/Symfony/Component/Form/Extension/Security/SecurityExtension.php:
```diff
@@ -0,0 +1,41 @@
+{
+    private $passwordEncoder;
+
+    public function __construct(UserPasswordEncoderInterface $passwordEncoder = null)
```
Please, remove the nullable definition.

yceruto commented:
I like the idea too! 

I'm wondering if this implementation fits better in a new "password type" extension rather than creating a new type and enable this feature through a new option "auto_encode".

I don't like much the idea to use the user instance. Instead, I would do what @wouterj has suggested here https://github.com/symfony/symfony/pull/35654#discussion_r376790165.

fabpot commented:
@pyrech @lyrixx Have you read the comments from @yceruto? Any interest in resuming the work here?

yceruto commented on src/Symfony/Component/Form/Extension/Security/Type/SecurityPasswordType.php:
```diff
@@ -0,0 +1,70 @@
+
+            $plainPassword = $event->getData();
+
+            $event->setData($plainPassword ? $this->passwordEncoder->encodePassword($securityUser, $plainPassword) : $securityUser->getPassword());
```
The submitted data cannot be changed this way before validating the plain password.

yceruto replied:
In effect, if the validation step fails, the encoded password will be visible to the end user.

yceruto reviewed (changes requested):
After a second review I realized that this password encoding cannot be done before validation (that happens in `POST_SUBMIT` event on the root form) otherwise constraint validators will validate the encoded password (which would be wrong) rather than the plain password. Let's think about the `Length()` constraint for instance.

But, also changing the form data in `POST_SUBMIT` event is currently impossible, so I don't think the proposal is feasible this way.

fabpot commented:
@pyrech Can you have a look at the feedback?

pyrech commented:
I agree with the reviews that the current implementation is not ideal at all. I don't know how to improve it enough though so we might just close this PR :+1:

Seb33300 commented:
I worked on a different approach here: https://github.com/symfony/symfony/pull/42807
